### PR TITLE
L3 Muon Filter Fix (Backport to 91X)

### PR DESCRIPTION
**Pull request by BenjaminRS**
This is a backport of PR ~~#18598~~ #18602

### CODE REVIEW

File: HLTrigger/Muon/src/HLTMuonTrimuonL3Filter.cc [hunk 1]
```diff
@@ -135,15 +139,55 @@ HLTMuonTrimuonL3Filter::hltFilter(edm::Event& iEvent, const edm::EventSetup& iSe
    Handle<RecoChargedCandidateCollection> mucands;
    if (saveTags()) filterproduct.addCollectionTag(candTag_);
    iEvent.getByToken(candToken_,mucands);
+
+   // Test to see if we can use L3MuonTrajectorySeeds:
+   if (mucands->size()<1) return false;
+   auto tk = (*mucands)[0].track();
+   bool useL3MTS=false;
```
Comment (fwyzard) [inline HLTrigger/Muon/src/HLTMuonTrimuonL3Filter.cc]: same comments as above

**Review by fwyzard — CHANGES_REQUESTED** [review]

File: HLTrigger/Muon/src/HLTMuonDimuonL3Filter.cc [hunk 2]
```diff
@@ -161,19 +165,63 @@ HLTMuonDimuonL3Filter::hltFilter(edm::Event& iEvent, const edm::EventSetup& iSet
    Handle<RecoChargedCandidateCollection> mucands;
    if (saveTags()) filterproduct.addCollectionTag(candTag_);
    iEvent.getByToken(candToken_,mucands);
+
+   // Test to see if we can use L3MuonTrajectorySeeds:
+   if (mucands->size()<1) return false;
+   auto tk = (*mucands)[0].track();
```
Comment (BenjaminRS) [inline HLTrigger/Muon/src/HLTMuonDimuonL3Filter.cc]: This is a tracker track; so maybe I can just dereference the track here:
https://github.com/BenjaminRS/cmssw/blob/9a5a3d38ba2cfe1bda5fb5598b46f39f6023d8d5/HLTrigger/Muon/src/HLTMuonDimuonL3Filter.cc#L175

File: HLTrigger/Muon/src/HLTMuonDimuonL3Filter.cc [hunk 3]
```diff
@@ -161,19 +165,63 @@ HLTMuonDimuonL3Filter::hltFilter(edm::Event& iEvent, const edm::EventSetup& iSet
...
+   bool useL3MTS=false;
+
+   if (tk->seedRef().isNonnull()){
+	   auto a = dynamic_cast<const L3MuonTrajectorySeed*>(tk->seedRef().get());
```
Comment (BenjaminRS) [inline HLTrigger/Muon/src/HLTMuonDimuonL3Filter.cc]: The dynamic_cast doesn't work if we are not using L3MuonTrajectorySeed ... if we dont use a L3MuonTrajectorySeed then we are using the normal TrajectorySeeds.

File: HLTrigger/Muon/src/HLTMuonDimuonL3Filter.cc (hunk 2 above)
Comment (fwyzard) [inline HLTrigger/Muon/src/HLTMuonDimuonL3Filter.cc]: I guess it is a TrackRef, not a track ?
Copying the TrackRef should be cheap; taking a `const &` should still be cheaper, though.

File: HLTrigger/Muon/src/HLTMuonDimuonL3Filter.cc (hunk 3 above)
Comment (fwyzard) [inline HLTrigger/Muon/src/HLTMuonDimuonL3Filter.cc]: ok

File: HLTrigger/Muon/src/HLTMuonDimuonL3Filter.cc [hunk 4]
```diff
@@ -161,19 +165,63 @@ HLTMuonDimuonL3Filter::hltFilter(edm::Event& iEvent, const edm::EventSetup& iSet
...
+   else{
+     // Read Links collection:
+     edm::Handle<reco::MuonTrackLinksCollection> links;
+     iEvent.getByToken(linkToken_, links);
+
+     // Loop over RecoChargedCandidates:
+     for(unsigned int i(0); i < mucands->size(); ++i){
+	RecoChargedCandidateRef cand(mucands,i);
+	for(auto const & l : *links){
+	  const reco::MuonTrackLinks* link = &l;
```
Comment (fwyzard) [inline HLTrigger/Muon/src/HLTMuonDimuonL3Filter.cc]: why not simply
```c++
for(auto const & link : *links){
  ...
  const reco::Track& globalTrack = * link.globalTrack();
```

etc ?

File: HLTrigger/Muon/src/HLTMuonDimuonL3Filter.cc (hunk 2 above)
Comment (BenjaminRS) [inline HLTrigger/Muon/src/HLTMuonDimuonL3Filter.cc]: Yes its a TrackRef, ok I am testing with the const ref here and other suggestions below.

**Comment by Martin-Grunewald** [comment]
please test

**Comment by Martin-Grunewald** [comment]
+1

**Comment by BenjaminRS** [comment]
Hello all, PR #18602 was merged for the master branch yesterday. Is there anything more that needs to be done for this backport to 91X?
Thanks,
Benjamin

**Comment by davidlange6** [comment]
+1

**Pull request merged**